rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -206,23 +227,32 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
      * record event if message is repeated
      */
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
-        Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
+        if (message.getMessageType() != MessageType.BLOCK_MESSAGE && message.getMessageType() != MessageType.TRANSACTIONS) {
+            return true;
+        }
 
-        boolean contains = receivedMessages.contains(encodedMessage);
+        long currentTime = System.currentTimeMillis();
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
-            }
+        synchronized (receivedPeerMessages) {
+            logger.trace("Cleaning {} messages from rlp queue", receivedPeerMessages.size());
+            receivedPeerMessages.entrySet()
+                    .stream()
+                    .filter(entry -> currentTime - entry.getValue() > RECEIVED_MESSAGES_CACHE_DURATION)
+                    .forEach(entry -> receivedPeerMessages.remove(entry.getKey()));
+        }
+
+        Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
+        ReceivedPeerMessageKey receivedPeerMessageKey = new ReceivedPeerMessageKey(sender.getPeerNodeID(), encodedMessage);
 
-        } else {
+        if (receivedPeerMessages.containsKey(receivedPeerMessageKey)) {
             reportEventToPeerScoring(sender, EventType.REPEATED_MESSAGE, "Received repeated message on {}, not added to the queue");
+            this.receivedPeerMessages.put(receivedPeerMessageKey, currentTime);
+            return false;
         }
 
-        return !contains;
+        this.receivedPeerMessages.put(receivedPeerMessageKey, currentTime);
```

Comment:
This line is duplicated, you can probably call it just before the `if (receivedPeerMessages.containsKey(receivedPeerMessageKey))`

Reply:
Sorry, explained myself wrong probably, I should've mentioned creating a variable storing if already contained before doing the put.